### PR TITLE
slack: Map attachments.

### PR DESCRIPTION
**Pull request by rheaparekh**
Currently, the uploaded files are being added in the dir:` uploads/<realm_id>/<attachment_id>/<file_name>.`
In zulip export data those are added in : `uploads/realm_id/<number>/<hash>/<file_name>`

### CODE REVIEW

File: zulip/integrations/slack/slackdata2zulipdata.py [hunk 1]
```diff
@@ -250,8 +251,8 @@ def sanitize_token(token):
         msgs = json.load(open(slack_dir + '/%s/%s' % (channel, json_name)))
         for msg in msgs:
             text = msg['text']
-            if "has joined the channel" in text:
-                continue
+            # TODO directly ignore noisy messages from msg['subtype']
```
Comment (rht) [inline zulip/integrations/slack/slackdata2zulipdata.py]: i.e. `if msg['subtype'] in ["channel_join", "channel_leave", ...]: continue`?

File: zulip/integrations/slack/slackdata2zulipdata.py [hunk 2]
```diff
@@ -384,8 +386,67 @@ def check_user_mention(text):
...
+    # To convert url in form https:\/\/files.slack.com\/files-pri\/..  to https://files.slack.com//files-pri//..
+    def parse_url(url):
+        url = url.replace("\/\/", "//")
+        url = url.replace("\/", "//")
+        return url
+
+    # Download from the slack url and save it to the folder zulipdata/uploads/realm_id/id/filename
+    def save_attachment(url, id, name):
+        url = parse_url(url)
+        response = requests.get(url, stream=True)
```
Comment (rht) [inline zulip/integrations/slack/slackdata2zulipdata.py]: This would work only if the Slack URL is world-readable though. Which I don't think is the case.

Comment (rht) [reply zulip/integrations/slack/slackdata2zulipdata.py]: Maybe there should be a `try` and print a warning message if the files can't be downloaded.

Comment (rht) [reply zulip/integrations/slack/slackdata2zulipdata.py]: The download works in Travis though. So I assume for public Slack instances, these are world-readable.

File: zulip/integrations/slack/slackdata2zulipdata.py (hunk 1 above)
Comment (rht) [reply zulip/integrations/slack/slackdata2zulipdata.py]: I will do this later.

File: zulip/integrations/slack/slackdata2zulipdata.py [hunk 3]
```diff
@@ -384,8 +386,67 @@ def check_user_mention(text):
...
+        with open(upload_dir + '/' + str(id) + '/' + name, 'wb') as output_file:
+            shutil.copyfileobj(response.raw, output_file)
+
+    # Get attachments by checking all messages
+    json_names = os.listdir(slack_dir + '/' + channel)
+    for json_name in json_names:
+        msgs = json.load(open(slack_dir + '/%s/%s' % (channel, json_name)))
+        attachment_id = 1
+        msg_id_count = 1
+        for msg in msgs:
```
Comment (rht) [inline zulip/integrations/slack/slackdata2zulipdata.py]: I think the implementation would be cleaner if the loop over ALL the messages happens only once, near L252 in the function `channelmessage2zerver_message_for_one_stream`. When the subtype is "file_share", do the `save_attachment` operation, and append to zerver_attachment.

File: zulip/integrations/slack/slackdata2zulipdata.py (hunk 1 above)
Comment (rht) [reply zulip/integrations/slack/slackdata2zulipdata.py]: Nvm, I will leave this to you; I am focusing on stage 2.

Comment (rheaparekh) [reply zulip/integrations/slack/slackdata2zulipdata.py]: Okay, I'll update this.

File: zulip/integrations/slack/slackdata2zulipdata.py (hunk 2 above)
Comment (rheaparekh) [reply zulip/integrations/slack/slackdata2zulipdata.py]: So I'll the ` try` here?

Comment (rht) [reply zulip/integrations/slack/slackdata2zulipdata.py]: Yes. If an exception happens due to permission denied, then just stop constructing the attachment, and let it be in TOODOO for PM stuff.

File: zulip/integrations/slack/slackdata2zulipdata.py (hunk 3 above)
Comment (rheaparekh) [reply zulip/integrations/slack/slackdata2zulipdata.py]: Right, I thought this too, but then even zerver_attachments would be a list of list of dict. ( list of dict of all attachments in a channel appended to the main list). The export data is just a list of dict of all the attachments. So what should I go ahead with?

**Comment by rheaparekh** [comment]
I'll make the changes after #8 has been merged

**Comment by rht** [comment]
@rheaparekh rebase reminder from rht/python-zulip-api slack. I just rebased from zulip/python-zulip-api master

**Comment by rht** [comment]
Merged!

**Pull request closed without merging**